File: app/src/main/java/net/hearnsoft/gensokyoradio/trd/service/WebSocketService.java
package net.hearnsoft.gensokyoradio.trd.service;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

import net.hearnsoft.gensokyoradio.trd.R;
import net.hearnsoft.gensokyoradio.trd.beans.NowPlayingBean;
import net.hearnsoft.gensokyoradio.trd.beans.SocketClientBeans;
import net.hearnsoft.gensokyoradio.trd.db.SongHistoryDbHelper;
import net.hearnsoft.gensokyoradio.trd.model.SongDataModel;
import net.hearnsoft.gensokyoradio.trd.utils.Constants;
import net.hearnsoft.gensokyoradio.trd.utils.GlobalTimer;
import net.hearnsoft.gensokyoradio.trd.utils.NullStringToEmptyAdapterFactory;
import net.hearnsoft.gensokyoradio.trd.utils.SettingsPrefUtils;
import net.hearnsoft.gensokyoradio.trd.utils.ViewModelUtils;
import net.hearnsoft.gensokyoradio.trd.ws.GRWebSocketClient;

import java.net.URI;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;

public class WebSocketService extends Service {
    private static final String TAG = WebSocketService.class.getSimpleName();
    private static WsServiceInterface wsInterface;
    private final ExecutorService signalThreadPool = Executors.newSingleThreadExecutor();
    private GRWebSocketClient wsClient;
    private Gson gson;
    private SongDataModel songDataModel;
    private Handler toastHandler;
    private URI uri;
    private ConnectivityManager.NetworkCallback networkCallback;
    private ConnectivityManager connectivityManager;
    private int clientId;
    private int recheck = 0;
    private boolean isRunningService = false;

    public static void setCallback(WsServiceInterface dataInterface) {
        wsInterface = dataInterface;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        uri = URI.create(Constants.WS_URL);
        toastHandler = new Handler(Looper.getMainLooper());
        // 获取全局ViewModel
        songDataModel = ViewModelUtils.getViewModel(getApplication(), SongDataModel.class);
        gson = new GsonBuilder()
                .disableHtmlEscaping()
                .setLenient()
                .serializeNulls()
                .setPrettyPrinting()
                .enableComplexMapKeySerialization()
                .create();
        initWebSocket();
        // 初始化网络监听
        connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        setupNetworkCallback();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // only start once
        if (!isRunningService) {
            isRunningService = true;
            signalThreadPool.submit(this::initConn);
        }
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private void setupNetworkCallback() {
        networkCallback = new ConnectivityManager.NetworkCallback() {
            @Override
            public void onAvailable(@NonNull Network network) {
                super.onAvailable(network);
                Log.d(TAG, "Network available");
                songDataModel.getNetworkStatus().postValue(true);

                if (wsClient == null || !wsClient.isOpen()) {
                    signalThreadPool.submit(() -> {
                        initWebSocket();
                        initConn();
                    });
                }

                toastHandler.post(() ->
                        Toast.makeText(getApplicationContext(),
                                getString(R.string.network_restored),
                                Toast.LENGTH_SHORT).show());
            }

            @Override
            public void onLost(@NonNull Network network) {
                super.onLost(network);
                Log.d(TAG, "Network lost");
                songDataModel.getNetworkStatus().postValue(false);

                closeWsClient();
                recheck = 0;

                toastHandler.post(() ->
                        Toast.makeText(getApplicationContext(),
                                getString(R.string.network_lost),
                                Toast.LENGTH_SHORT).show());
            }
        };

        // 注册网络回调
        connectivityManager.registerDefaultNetworkCallback(networkCallback);
    }

    private void initWebSocket() {
        wsClient = new GRWebSocketClient(uri) {
            @Override
            public void onMessage(String message) {
                extractData(message);
            }

            @Override
            public void onClose(int code, String reason, boolean remote) {
                super.onClose(code, reason, remote);
                String reasonLog;
                if (code == 1000 && remote) {
                    reasonLog = "Socket connection closed cleanly,";
                } else {
                    reasonLog = "Socket connection closed unexpectedly, the connection will be retried at a later time.";
                    reConnectSocket();
                }
                Log.d(TAG, reasonLog + " code=" + code + ", reason=" + reason);
            }
        };
        wsClient.setSocketFactory(getDefaultSSLSocketFactory());
    }

    private void reConnectSocket() {
        recheck++;
        postSocketErrorToast(recheck);
        double timeOut;
        switch (recheck) {
            case 1:
                timeOut = Math.floor((Math.random() * 10) + 1);
                break;
            case 2:
                timeOut = Math.floor((Math.random() * 10) + 10);
                break;
            case 3:
                timeOut = Math.floor((Math.random() * 15) + 15);
                break;
            case 4:
                timeOut = Math.floor((Math.random() * 15) + 30);
                break;
            default:
                timeOut = Math.floor((Math.random() * 20) + 45);
                break;
        }
        Timer reConnTimer = new Timer();
        reConnTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                initWebSocket();
                initConn();
            }
        }, (long)(timeOut * 1000));
        Log.d(TAG, "will reconnect in " + (timeOut * 1000) + " millisecond later...");
    }

    /**
     * Initiate received data processing
     * 初始化接收到的数据处理
     * @param message
     */
    private void extractData(String message) {
        if (message != null) {
            Log.d(TAG, "socket message:" + message);
            if (isJson(message)) {
                if (message.contains("welcome")) {
                    SocketClientBeans clientBeans = gson.fromJson(message, SocketClientBeans.class);
                    clientId = clientBeans.id;
                    Log.d(TAG, "get Client ID: " + clientId);
                    SettingsPrefUtils.getInstance(this).writeIntSettings("clientId", clientId);
                } else if (message.equals("{\"message\":\"ping\"}")) {
                    Log.d(TAG, "get ping! send pong!");
                    sendPong();
                } else {
                    // Generate bean data
                    genBeanData(message);
                }
            } else if (message.startsWith("Error")) {
                toastHandler.post(() ->
                        Toast.makeText(getApplicationContext(),
                                getString(R.string.socket_error_server_msg)
                                + message, Toast.LENGTH_SHORT).show());
            } else {
                Log.e(TAG, "get invalid json data!");
                toastHandler.post(() ->
                        Toast.makeText(getApplicationContext(),
                                getString(R.string.socket_error_json_data),
                                Toast.LENGTH_SHORT).show());
            }
        } else {
            Log.e(TAG, "get null data!");
        }
    }

    /**
     * Get default SSLSocketFactory
     * 获取默认 SSLSocketFactory
     */
    private SSLSocketFactory getDefaultSSLSocketFactory() {
        SSLContext sslContext = null;
        try {
            sslContext = SSLContext.getDefault();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return sslContext.getSocketFactory();
    }

    /**
     * Send initial connection message
     * 发送初始连接消息
     * @params null
     */
    private void initConn() {
        // Connect to WebSocket server and waiting for welcome message
        // 连接到 WebSocket 服务器并等待 welcome 消息
        try {
            wsClient.connectBlocking();
        } catch (InterruptedException e) {
            Log.e(TAG, "webSocket connect time out!");
            e.printStackTrace();
        }
        // If the connection is successful, send the initial connection message
        // 如果连接成功，发送初始连接消息
        if (wsClient != null && wsClient.isOpen()) {
            wsClient.send(Constants.WS_SESSION_MSG);
        }
    }

    /**
     * Generate bean data
     * 从 json 反序列化生成 bean 数据
     * @param jsonData
     */
    private void genBeanData(String jsonData) {
        Gson gson = new GsonBuilder()
                .disableHtmlEscaping()
                .setLenient()
                .serializeNulls()
                .setPrettyPrinting()
                .enableComplexMapKeySerialization()
                .registerTypeAdapterFactory(new NullStringToEmptyAdapterFactory<String>())
                .create();
        NowPlayingBean bean = gson.fromJson(jsonData, NowPlayingBean.class);
        wsInterface.beanReceived(bean);
        //重置计数器并重新计时
        GlobalTimer.getInstance().startTimer(
                bean.getDuration(),
                bean.getPlayed()+1,
                bean.getRemaining()-1
        );
        long rowId = SongHistoryDbHelper.getInstance(this).insertSong(
                bean.getSongId(),
                bean.getTitle(),
                bean.getArtist(),
                bean.getAlbum(),
                bean.getCircle(),
                bean.getAlbumArt(),
                bean.getAlbumid()
        );
        Log.d(TAG, "record history row id "+ rowId);

        //genMediaNotification(bean);
        setViewModelData(bean);
        sendUpdateIntent();
    }

    private void sendUpdateIntent() {
        Intent updateIntent = new Intent();
        updateIntent.setAction("net.hearnsoft.gensokyoradio.trd.UPDATE_NOTIFICATION");
        LocalBroadcastManager.getInstance(this).sendBroadcast(updateIntent);
    }

    private void setViewModelData(NowPlayingBean bean) {
        songDataModel.getTitle().postValue(bean.getTitle());
        songDataModel.getArtist().postValue(bean.getArtist());
        songDataModel.getAlbum().postValue(bean.getAlbum());
        songDataModel.getCoverUrl().postValue(bean.getAlbumArt());
        songDataModel.getIsUpdatedInfo().postValue(true);
    }

    /**
     * Check data is JSON data
     * 检查数据是否为 JSON 数据
     * @param jsonData
     * @return boolean isJson
     */
    private boolean isJson(String jsonData) {
        JsonElement jsonElement;
        try {
            jsonElement = JsonParser.parseString(jsonData);
            return jsonElement.isJsonObject();
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Send pong message
     * 发送 pong 心跳包消息
     * @params null
     */
    private void sendPong() {
        wsClient.send("{\"message\":\"pong\", \"id\":" + clientId + "}");
    }

    /**
     * 显示WebSocket错误断开toast
     */
    private void postSocketErrorToast(int recheck) {
        toastHandler.post(() ->
                Toast.makeText(getApplicationContext(),
                        getString(R.string.socket_error_reconnect) +
                                recheck, Toast.LENGTH_SHORT).show());
    }

    /**
     * Close WebSocket client
     * 关闭 WebSocket 客户端
     * @params null
     */
    private void closeWsClient() {
        try {
            if (wsClient != null) {
                wsClient.close();
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            wsClient = null;
        }
    }

    @Override
    public void onDestroy() {
        // 注销网络监听
        if (connectivityManager != null && networkCallback != null) {
            try {
                connectivityManager.unregisterNetworkCallback(networkCallback);
            } catch (Exception e) {
                Log.e(TAG, "Failed to unregister network callback", e);
            }
        }

        super.onDestroy();
        closeWsClient();
        isRunningService = false;
    }

}

File: app/src/main/java/net/hearnsoft/gr3rd/compose/service/WebSocketService.kt
package net.hearnsoft.gr3rd.compose.service

import android.app.Service
import android.content.Intent
import android.net.ConnectivityManager
import android.net.Network
import android.widget.Toast
import com.blankj.utilcode.util.SPStaticUtils
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.Strictness
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.hearnsoft.gr3rd.compose.R
import net.hearnsoft.gr3rd.compose.domain.beans.NowPlayingData
import net.hearnsoft.gr3rd.compose.domain.beans.RadioClientData
import net.hearnsoft.gr3rd.compose.domain.viewmodel.SongViewModel
import net.hearnsoft.gr3rd.compose.socket.GRWebSocketClient
import net.hearnsoft.gr3rd.compose.utils.Constants
import net.hearnsoft.gr3rd.compose.utils.Logger
import net.hearnsoft.gr3rd.compose.utils.NullStringToEmptyAdapterFactory
import java.net.URI
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLSocketFactory
import kotlin.random.Random

class WebSocketService : Service() {
    companion object {
        const val TAG = "WebSocketService"
    }

    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // 使用共享的 ViewModel 实例
    private val songViewModel by lazy { SongViewModel.getInstance() }

    private lateinit var uri: URI
    private lateinit var gson: Gson
    private lateinit var connMgr: ConnectivityManager
    private lateinit var webSocketClient: GRWebSocketClient
    private lateinit var networkCallback: ConnectivityManager.NetworkCallback

    private var isRunning = false
    private var recheck = 0
    private var clientId: Int = 0

    private val pong: JsonObject = JsonObject().apply {
        addProperty("message", "pong")
        addProperty("id", clientId)
    }

    override fun onCreate() {
        super.onCreate()
        // 在 onCreate 中初始化
        uri = URI.create(Constants.WS_URL)
        gson = GsonBuilder()
            .disableHtmlEscaping()
            .setStrictness(Strictness.LENIENT)
            .serializeNulls()
            .setPrettyPrinting()
            .enableComplexMapKeySerialization()
            .registerTypeAdapterFactory(NullStringToEmptyAdapterFactory())
            .create()

        // 初始化 WebSocket 客户端
        initWebSocketClient()

        // 初始化网络监听
        connMgr = getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
        setupNetworkCallback()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        // 只启动一次
        if (!isRunning) {
            isRunning = true
            serviceScope.launch {
                initWSConn()
            }
        }
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onBind(intent: Intent?) = null

    // 设置网络回调
    private fun setupNetworkCallback() {
        networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                super.onAvailable(network)
                Logger.info(TAG, "Network is available")

                // 更新ViewModel网络状态、
                songViewModel.updateNetworkStatus(true)

                if (!webSocketClient.isOpen) {
                    Logger.info(TAG, "Network available, attempting to reconnect WebSocket...")
                    serviceScope.launch {
                        initWebSocketClient()
                        initWSConn()
                    }
                }

                serviceScope.launch {
                    showNetworkStatusToast(true)
                }
            }

            override fun onLost(network: Network) {
                super.onLost(network)
                Logger.warn(TAG, "Network is lost")

                // 更新 ViewModel 中的网络状态
                songViewModel.updateNetworkStatus(false)
                songViewModel.updateWebSocketStatus(false)

                closeWsClient()

                serviceScope.launch {
                    showNetworkStatusToast(false)
                }
            }
        }
        // 注册网络回调
        connMgr.registerDefaultNetworkCallback(networkCallback)
    }

    // 初始化 WebSocket 客户端
    private fun initWebSocketClient() {
        webSocketClient = GRWebSocketClient(
            serverUri = uri,
            onMessage = {
                serviceScope.launch {
                    extractData(it)
                }
            },
            onOpen = {
                // 连接成功时更新状态
                Logger.info(TAG, "WebSocket connected successfully")
                songViewModel.updateWebSocketStatus(true)
                recheck = 0 // 重置重连计数
            },
            onClose = { code, reason, remote ->
                Logger.info(TAG, "WebSocket connection closed: code=${code}, reason=${reason}, remote=${remote}")
                songViewModel.updateWebSocketStatus(false)
                // 如果是正常关闭（code 1000），则不需要重连
                if (code != 1000 || !remote) {
                    // 处理非正常关闭
                    Logger.warn(TAG, "WebSocket closed unexpectedly: code=$code, reason=$reason")
                    Logger.info(TAG, "Attempting to reconnect WebSocket...")
                    // 重新连接 WebSocket
                    serviceScope.launch {
                        reconnectWebSocket()
                    }
                }
            },
            onError = {
                Logger.err(TAG, "WebSocket error occurred", it)
                songViewModel.updateWebSocketStatus(false)
            }
        )
        webSocketClient.setSocketFactory(getDefaultSSLSocketFactory())
    }

    // 初始化 WebSocket 连接
    private fun initWSConn() {
        try {
            webSocketClient.connectBlocking()
            // 连接成功后发送欢迎消息并更新状态
            if (webSocketClient.isOpen) {
                // 连接成功后，发送欢迎消息
                Logger.info(TAG, "WebSocket client initialized and connected")
                songViewModel.updateWebSocketStatus(true)
                webSocketClient.send(Constants.WS_SESSION_MSG.toString())
            }
        } catch (e: InterruptedException) {
            // 处理异常
            e.printStackTrace()
        } catch (e: Exception) {
            // 处理其他异常
            e.printStackTrace()
        }
    }

    // 处理接收到的数据
    private suspend fun extractData(data: String) {
        if (data.isEmpty()) {
            Logger.warn(TAG, "Received empty data")
            return
        }

        data.let {data ->
            Logger.debug(TAG, "Received data: $data")

            when {
                isJson(data) -> handleJsonData(data)
                data.startsWith("Error") -> showErrorToast(data)
                else -> Logger.warn(TAG, "Unknown data format: $data")
            }
        }
    }

    // 处理 JSON 数据
    private fun handleJsonData(data: String) {
        when {
            data.contains("welcome") -> {
                val clientData = gson.fromJson(data, RadioClientData::class.java)
                clientId = clientData.id
                Logger.info(TAG, "WebSocket connected with client ID: ${clientData.id}")
                SPStaticUtils.put(Constants.PREF_CLIENT_ID, clientId)
            }
            data == Constants.WS_PING_MSG.toString() -> {
                Logger.debug(TAG, "Received ping from server, sending pong")
                sendPong()
            }
            else -> genBeanData(data)
        }
    }

    // 处理服务器发来的专辑信息
    private fun genBeanData(string: String) {
        try {
            val nowPlayingData = gson.fromJson(string, NowPlayingData::class.java)
            // 更新 ViewModel 中的当前播放数据
            songViewModel.updateFromWebSocket(nowPlayingData)
            Logger.info(TAG, "Updated now playing data: ${nowPlayingData.title}")
        } catch (e: Exception) {
            Logger.err(TAG, "Error parsing now playing data", e)
        }
    }

    // 关闭 WebSocket 客户端
    private fun closeWsClient() {
        try {
            if (webSocketClient.isOpen) {
                webSocketClient.close()
            }
        } catch (e: Exception) {
            Logger.err(TAG, "Error closing WebSocket client", e)
        } finally {
            webSocketClient.close()
            isRunning = false
            recheck = 0
        }
    }

    // 重新连接 WebSocket
    private suspend fun reconnectWebSocket() {
        recheck++
        postSocketErrorToast(recheck)

        val delayMillis = calculateReconnectDelay(recheck)
        Logger.info(TAG, "Will reconnect in ${delayMillis}ms...")

        delay(delayMillis)

        try {
            initWebSocketClient()
            initWSConn()
        } catch (e: Exception) {
            Logger.err(TAG, "Reconnection failed", e)
        }
    }

    // 发送心跳包
    private fun sendPong() {
        if (webSocketClient.isOpen) {
            webSocketClient.send(pong.toString())
            Logger.debug(TAG, "Sent ping to WebSocket server")
        } else {
            Logger.warn(TAG, "WebSocket is not open, cannot send ping")
        }
    }

    // 计算重连延时
    private fun calculateReconnectDelay(retryCount: Int): Long {
        val baseDelay = when (retryCount) {
            1 -> 1_000L to 10_000L
            2 -> 10_000L to 20_000L
            3 -> 15_000L to 30_000L
            4 -> 30_000L to 45_000L
            else -> 45_000L to 65_000L
        }
        return Random.nextLong(baseDelay.first, baseDelay.second)
    }

    private fun getDefaultSSLSocketFactory() : SSLSocketFactory {
        return try {
            val sslContext = SSLContext.getDefault()
            sslContext.init(null, null, null)
            sslContext.socketFactory
        } catch (e: Exception) {
            Logger.err(TAG, "Error creating SSLContext", e)
            SSLContext.getDefault().socketFactory
        }
    }

    private suspend fun postSocketErrorToast(recheckCount: Int) {
        withContext(Dispatchers.Main) {
            Toast.makeText(
                applicationContext,
                getString(R.string.socket_error_reconnect) + recheckCount,
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    private suspend fun showErrorToast(message: String) {
        withContext(Dispatchers.Main) {
            Toast.makeText(
                applicationContext,
                getString(R.string.socket_error_server_msg) + message,
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    private fun isJson(data: String) : Boolean {
        val jsonElement = JsonParser.parseString(data);
        return jsonElement.isJsonObject;
    }

    private suspend fun showNetworkStatusToast(isAvailable: Boolean) {
        val message = if (isAvailable) {
            getString(R.string.network_restored)
        } else {
            getString(R.string.network_lost)
        }
        withContext(Dispatchers.Main) {
            Toast.makeText(
                applicationContext,
                message,
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        closeWsClient()
        isRunning = false
    }
}